nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala

```diff
@@ -31,381 +36,698 @@ sealed abstract class Array[T]
 
   /** Create a shallow of given array. */
   protected override def clone(): Array[T] = undefined
-}
-
-final class ObjectArray private () extends Array[Object] {
-  def apply(i: Int): Object =
+  
+  /*
+   * helper methods
+   */
+   
+  /** called by subclass::apply and subclass::update */
+  protected def pointerAt(sizeOneElement: CSize, i: Int): Ptr[_] = {
     if (i < 0 || i >= length)
       throw new IndexOutOfBoundsException(i.toString)
     else {
-      val headptr = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader]).cast[Ptr[Object]]
-      headptr(i)
+      (this.cast[Ptr[Byte]] + sizeof[ArrayHeader] + sizeOneElement * i).cast[Ptr[_]]
     }
+  }
+  
+  /** called by subclass::clone.  Allocate memory and copy header and data */
+  protected def helperClone(sizeofElement: CSize): Ptr[_] = {
+    val arrsize = sizeof[ArrayHeader] + sizeofElement * length
+    val arr = runtime.alloc(arrsize)
+    `llvm.memcpy.p0i8.p0i8.i64`(arr.cast[Ptr[Byte]], this.cast[Ptr[Byte]], arrsize, 1, false)
+    arr        
+  }  
+}
 
-  def update(i: Int, value: Object): Unit =
-    if (i < 0 || i >= length)
-      throw new IndexOutOfBoundsException(i.toString)
-    else {
-      val headptr = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader]).cast[Ptr[Object]]
-      headptr(i) = value
+/** Utilities methods
+  *
+  */
+
+object Array {
+  /*
+   *
+   */    
+  def copy (from: AnyRef, fromPos: Int, to: AnyRef, toPos: Int, length: Int): Unit = {
+    if (from == null)
+      throw new NullPointerException()
+    
+    if (to == null)
+      throw new NullPointerException()
+    
+    val fromTypeId = instanceTypeId(from)
+    val toTypeId = instanceTypeId(to)
+    
+    if (fromTypeId == arrayObjectTypeId && toTypeId == arrayObjectTypeId) 
+      throw new ArrayStoreException ("copy from object array to object array isn't implemented yet")    
```

Comment:
Object arrays should have the same implementation as we have now for primitives. We won't check safety of covariant casts, but that's fine for now, nobody does that anyway.